### PR TITLE
When byteorder record is missing load as little endian by default

### LINKED ISSUE #101688: torch.load() may store unexpected values into tensors on big-endian
### 🐛 Describe the bug

The current `torch.load()` interprets data as native-endian if a given pickle file has no `byteorder` record.

This behavior causes practical issues on big-endian machines when we load parameters for many DNN models saved on little-endian machines (e.g. x86_64). As a result, tensors would have incorrect values. Practically, it would be better to handle the pickle file w/o `byteorder` record as little-endian. 

In addition, we could add an option to handle the pickle file w/o `byteorder` record as native-endian to keep the current behavior.

```python
import torch

x = torch.load(`path to pickle file w/o byteorder record`)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0a0+gite7681b5
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (s390x)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-71-generic-s390x-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False

CPU:
Architecture:                    s390x
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Big Endian
CPU(s):                          8
On-line CPU(s) list:             0-7
Vendor ID:                       IBM/S390
Machine type:                    3931
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s) per book:              1
Book(s) per drawer:              1
Drawer(s):                       8
CPU dynamic MHz:                 5200
CPU static MHz:                  5200
BogoMIPS:                        26315.00
Dispatching mode:                horizontal
Flags:                           esan3 zarch stfle msa ldisp eimm dfp edat etf3eh highgprs te vx vxd vxe gs vxe2 vxp sort dflt vxp2 nnpa sie
Hypervisor:                      z/VM 7.2.0
Hypervisor vendor:               IBM
Virtualization type:             full
L1d cache:                       1 MiB (8 instances)
L1i cache:                       1 MiB (8 instances)
L2 cache:                        256 MiB (8 instances)
L3 cache:                        256 MiB
NUMA node(s):                    1
NUMA node0 CPU(s):               0-7
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Not affected
Vulnerability Spectre v1:        Mitigation; __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; etokens
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0a0+gite7681b5
[conda] Could not collect
```

### PR DESCRIPTION
**Pull request by AlekseiNikiforovIBM**
Fixes #101688

### CODE REVIEW

**Review by mikaylagawarecki — APPROVED** [review]
Following discussion here, I think it makes sense to approve this change and cherry pick it into the release branch as a critical fix to a new feature https://github.com/pytorch/pytorch/issues/101688#issuecomment-1701308630

**Review by mikaylagawarecki — COMMENTED** [review]
Could you perhaps add a small test to make sure the behavior/warning is raised as expected

**Comment by AlekseiNikiforovIBM** [comment]
Sure. Are there any tests for warnings I could use as reference?

File: test/test_serialization.py [hunk 1]
```diff
@@ -3859,6 +3860,53 @@ def test_serialization_load_bom_data_cfloat(self):
...
+                          b'\x00\x00\x00\x00\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\xd2\x00\x00\x00\x00'
+                          b'\x00\x00\x00\xd2\x01\x00\x00\x00\x00\x00\x00PK\x06\x07\x00\x00\x00\x00\xa4\x02\x00'
+                          b'\x00\x00\x00\x00\x00\x01\x00\x00\x00PK\x05\x06\x00\x00\x00\x00\x03\x00\x03\x00'
+                          b'\xd2\x00\x00\x00\xd2\x01\x00\x00\x00\x00')
+
+        current_load_endian = get_default_load_endianness()
+
+        buf_be_no_bom = io.BytesIO(data_be_no_bom)
+
+        with warnings.catch_warnings(record=True) as warns:
```
Comment (mikaylagawarecki) [inline test/test_serialization.py:3895]: tiny nit: Here is an example of of how we usually catch warnings in test

https://github.com/pytorch/pytorch/blob/0f1a225f33b6e3966eda02baec5154fdc39a18dd/test/test_nn.py#L2703

**Review by mikaylagawarecki — COMMENTED** [review]
Feel free to merge after you fix the nits and CI is green :)

After this is merged could you please open a PR to cherry pick this into the release branch. See instructions here
https://github.com/pytorch/pytorch/issues/108055. The criteria category should be "Critical fixes to new feature"

File: torch/serialization.py [hunk 2]
```diff
@@ -1326,27 +1326,27 @@ def _load(zip_file, map_location, pickle_module, pickle_file='data.pkl', overall
...
         # See https://github.com/pytorch/pytorch/issues/101688
         warnings.warn("The default load endianness for checkpoints without a byteorder mark "
-                      "on big endian machines will be changed from 'native' to 'little' endian "
-                      "in a future release, to avoid this behavior please use "
+                      "on big endian machines was changed from 'native' to 'little' endian, "
+                      "to avoid this behavior please use "
                       "torch.serialization.set_default_load_endianness to set "
                       "the desired default load endianness",
-                      DeprecationWarning)
+                      FutureWarning)
```
Comment (mikaylagawarecki) [inline torch/serialization.py:1349]: nit: `UserWarning`

**Comment by mikaylagawarecki** [comment]
Looks like lint is failing

To fix the failing lint job running `make setup_lint` and `make quicklint` from your local pytorch directory and following the corresponding instructions should be sufficient, see [here](https://github.com/pytorch/pytorch/blob/4721553431e968ee5e90753346d92252cce98e84/Makefile#L25-L34)

**Pull request closed without merging**